expression/expression.go

```diff
@@ -309,6 +328,96 @@ func VecEvalBool(ctx sessionctx.Context, exprList CNFExprs, input *chunk.Chunk,
 	return selected, nulls, nil
 }
 
+func toBool(sc *stmtctx.StatementContext, eType types.EvalType, buf *chunk.Column, sel []int, isZero []int8) error {
+	var err error
+	switch eType {
+	case types.ETInt:
+		i64s := buf.Int64s()
+		for i := range sel {
+			if buf.IsNull(i) {
+				isZero[i] = -1
+			} else {
+				if i64s[i] == 0 {
+					isZero[i] = 0
+				} else {
+					isZero[i] = 1
+				}
+			}
+		}
+	case types.ETReal:
+		f64s := buf.Float64s()
+		for i := range sel {
+			if buf.IsNull(i) {
+				isZero[i] = -1
+			} else {
+				if types.RoundFloat(f64s[i]) == 0 {
```

Comment:
I'm not sure whether we need to Round here.
``` sql
mysql> select 0.1e1 && 1;
+------------+
| 0.1e1 && 1 |
+------------+
|          1 |
+------------+
1 row in set (0.00 sec)

mysql> select 0.1 && 1;
+----------+
| 0.1 && 1 |
+----------+
|        1 |
+----------+
1 row in set (0.00 sec)

```

Reply:
In keeping with previous implementations. See [here](https://github.com/pingcap/tidb/blob/master/expression/expression.go#L198) and [here](https://github.com/pingcap/tidb/blob/master/types/datum.go#L1364)

Reply:
I know it's the same as the previous implementation.
What I mean here is that we may confirm whether this function is not correct and we need to refine it later.

Reply:
@XuHuaiyu 
In mysql 5.7, the logical result in expression is different from result in where clause. Here's an example.

```mysql
mysql> select 1 and 0.5;
+-----------+
| 1 and 0.5 |
+-----------+
|         1 |
+-----------+
1 row in set (0.00 sec)

mysql> select 1 and 0.49;
+------------+
| 1 and 0.49 |
+------------+
|          1 |
+------------+
1 row in set (0.01 sec)

mysql> select * from t where 1 and 0.5;
+------+
| c1   |
+------+
|    9 |
+------+
1 row in set (0.00 sec)

mysql> select * from t where 1 and 0.49;
Empty set (0.00 sec)
```

Reply:
`ToBool` evaluate the logical result in where clause, and it seems to be compatible with mysql5.7 for now.

The logical result in expression will be fixed in https://github.com/pingcap/tidb/pull/12173.

Reply:
@Reminiscent we may add these examples into unit test?

Reply:
@Reminiscent I think we'd better add some tests for the cases that @SunRunAway mentioned.